### ISSUE TITLE
[Feature] Changing statefulness of RNN between training and prediction stages

### ISSUE THREAD

**Issue by joaospinto** [fchollet/keras#7441]
Hi everyone,
I hope this is an appropriate place for my question/proposal - please close this otherwise.
I've started using Keras recently at work, and for my particular use-case it seems to make more sense to use stateless RNNs during the training stage, essentially because the sequences I feed into my network are of an episodic nature; when a sequence is over, sharing its state to the first element of the next sequence would not make much sense and would introduce noise. Therefore, I implemented a generator for batches, which makes sure that separate episodes are never mixed together.
However, in the prediction stage, there is a single continuous episode (I guess because I'm using this network for a real-time problem); my intuition is that this is quite a common setting, rather than a project-specific one. Since I made my RNN stateless, I need to buffer _lookahead_-many inputs into a 3-dimensional numpy array with dimension (_batchsize_=1, _lookahead_, _inputdim_). It seems to me that it would be quite useful to have an option to change the statefulness of the RNN between these two stages (if nothing else is feasible, maybe by simply padding the tensor that is fed into the network). Would that make sense, or is it too specific a feature? I would be glad to help with it if it does.
Regards

**Comment by td2014** [comment]
@joaospinto .  I'm trying to understand your use case a little bit better.  Your training data is episodic, but the inference data is continuous (real-time as you say).  What I'm wondering about is if your training data and inference data don't have similar temporal characteristics, is your trained model valid?  In other words, if you are able to represent your data as episodic and consider it valid, shouldn't the same pre-processing apply to the real-time data.  Perhaps I'm missing a subtlety, but I typically think of the training data having to be representative of the actual.  I hope I'm understanding correctly - just trying to get some clarification of your challenge.  Thanks.

**Comment by joaospinto** [comment]
Hi @td2014, thanks for your answer and reasonable questions. Allow me to expand a little more, in this case. I'm using the network to learn how to control a robot, using a previous controller or a human as a teacher. The training-stage episodes consist of separate runs, and at prediction stage there is only the run that the robot is taking at that time. Furthermore, at training stage, no input to the network refers to either a different episode or a future time, so there is indeed a consistency between both stages. I hope this helps - if not, please let me know and I can expand further. Thank you for helping.

**Comment by td2014** [comment]
@joaospinto .  Thanks for the additional information.  Perhaps I'm still not fully understanding, but it seems that if you are training your network without retaining network state information from batch to batch (for the relevant example in each batch), then for the prediction mode, it doesn't seem to make sense to me why retaining network state would be needed.  If you present the current robot agent state and environment observations as inputs to the RNN (like you are doing in training), the predicted output result should corresponding to what you need to generate the next robot action, if I'm understanding your method correctly.  The following documentation may also help:

From the Keras FAQ: https://keras.io/getting-started/faq/#how-can-i-use-stateful-rnns
 "How can I use stateful RNNs?
Making a RNN stateful means that the states for the samples of each batch will be reused as initial states for the samples in the next batch."

It sounds like you have an interesting project, so I hope this helps.  Thanks.

**Comment by joaospinto** [comment]
Thanks your help @td2014 . As you said, during training time I don't retain any state across batches, because the way I select the 3rd-order tensor is by selecting each component of the batch independently (selecting an episode uniformly at random), and selecting a contiguous segment of size _timesteps_ within that episode. Because of the way this is done, it would not make sense for me to carry over state to the next generated batch. At evaluation stage, I can change my batch size to 1 (since I'm evaluating in real-time), but I still need to give _timesteps_-many examples, each of size _inputdim_, which I need to somehow buffer. I guess the difference is that in the prediction stage there is a temporal order across batches, which does not exist during training. Does what I'm doing make sense? If the network were stateful at prediction time, my perception (do correct me if my understanding of how it works is flawed) is that this would prevent me from having to buffer/pad the tensor that I feed into _model.predict_ myself. Cheers!

**Comment by td2014** [comment]
@joaospinto .  Thanks for the details.  I may still be missing the subtleties, but let me try to explain my understanding and see if I can connect it to your situation.  The RNN is always "stateful" in some sense, because in addition to the weights which have been trained, the network has an internal state which is propagated from timestep to timestep within a given example in order to make the recursive calculations which depend on the output of the previous timestep etc.  The batchsize is not really a factor, only the number of timesteps (and the input dim per timestep).  During training, a batch can have many examples of input sequences (num timesteps x input dim).  When you are predicting using the model, you are really only providing a single example (as you mentioned) but the number of timesteps (and input dim) should be consistent with the number of timesteps you used in the training (not the batch size).

For example, if we had a batch size of 3, each with 5 time steps, and 2 input dims, our training data would look like this:

example1:  timestep 1    timestep2   timestep3   timestep4   timestep5
                   inputval1       inputval1     inputval1     inputval1     inputval1
                   inputval2      inputval2     inputval2     inputval2    inputval2
example2:  timestep 1    timestep2   timestep3   timestep4   timestep5
                   inputval1       inputval1     inputval1     inputval1     inputval1
                   inputval2      inputval2     inputval2     inputval2    inputval2
example3:  timestep 1    timestep2   timestep3   timestep4   timestep5
                   inputval1       inputval1     inputval1     inputval1     inputval1
                   inputval2      inputval2     inputval2     inputval2    inputval2

But during prediction, you only want:
prediction1:  timestep 1    timestep2   timestep3   timestep4   timestep5
                   inputval1       inputval1     inputval1     inputval1     inputval1
                   inputval2      inputval2     inputval2     inputval2    inputval2


So, as I was mentioning earlier, the key is that the number of timesteps match in the training and prediction stages.  At the end of 5 timesteps in this case, you would generate your first prediction etc.
This is how I am understanding the process.  I hope it makes sense.  Thanks.

**Comment by joaospinto** [comment]
@td2014 Thanks for expanding in such detail. As you said, since I must make the number of timesteps consistent between the training and prediction stages, this creates a need buffering them (at least in application cases like mine, where prediction is done in real-time); this can be done fairly efficiently, by keeping a numpy array of shape (_bufferratio_ * _timesteps_, _inputdim_) and clearing the buffer every _bufferratio_ predictions (and always returning the right slice of this numpy array). However, if we were not erasing the state of the network after each prediction (even though the network was trained as stateless), would this not be a significant gain in computational efficiency? Essentially because we are always calling it with an array of timesteps that is always right-shifted by one in the prediction stage. Perhaps my use case corresponds to only a minority of the applications of RNNs, but I could see some sense in having the default behaviour of _model.predict_ be keeping the state or not as a function of the shape of the given tensor (in this case, if it saw a 2-nd order tensor, it would keep the state, and if it saw a 3-rd order tensor would just unroll the network similarly to what happens in training and evaluate it in the corresponding fashion).

**Comment by td2014** [comment]
@joaospinto .  Thanks for your additional comments.  Just to make sure we are speaking about the same concept of "state" can you please describe what you mean when you say "state of the network?"  In other words, what parameters of the network are you considering as "state"?  This might help me understand better your perspective.  When I am saying state, I mean what ever what the output of the network after the final timestep (including any "internal memory" like in the case of the LSTM cell).  I don't include "weights" when I say "state."  The weights are fixed during prediction, and therefore cannot be updated (whereas the state, the way I have defined it, can).  Thanks again for discussing your project and improvement ideas.  I hope this discussion helps.  Thanks.

**Comment by joaospinto** [comment]
@td2014 Thanks again for replying. I think we are using the same definition of "state" (which is obviously different from the use of the word "stateful" in Keras, which I suppose corresponds to whether the RNN is unrolled or not); in particular, I don't include weights when I say state, but rather the outputs of the network (more precisely, the h_t's in [this image](http://colah.github.io/posts/2015-08-Understanding-LSTMs/img/LSTM3-var-GRU.png); the image depicts a GRU, which is actually what I am using, not that it makes much difference for us; as you said, other "internal memory" would also be included in my definition of state). Essentially, my understanding is that if we declare the network to be stateless then all the h_t will be erased after each training/prediction step, and if we declare it to be stateful then the last h_t is kept (the prior ones can be removed since they are no longer useful, of course). I suppose the remark that I'm trying to make (although, of course, I might be missing something) is that in settings where one uses the network for real-time prediction, even though the network might have been trained in stateless mode, having a mode for stateful prediction would dispense the need to buffer/pad the lasts _timesteps_-many prediction inputs. Thank you so much for helping me clarify this.

**Comment by casparcaspar** [comment]
@joaospinto I see exactly your point and I'm actually facing the same issue at the right now. My RNN-model is supposed to predict in realtime and therefore I prefer the stateful mode since (just as you mentioned) otherwise I would have to keep track of previous inputs. I also assume that a stateless model performs a simulation on the input sequence for each prediction. In my understanding a single prediction step with a stateful model works more like a feedforward net execution since the state is available to the model from the last prediction. So why don't I want to use stateful mode for training? Because I don't have one long training sequence but many short ones, which is why a state reset between using each for training would be good. Also for my application long term dependencies aren't really important. So if you've come to a solution or a workaround to our problem I would be extremely interested to read about it. 😀

**Comment by JustAnotherUsername123** [comment]
@joaospinto and @casparcaspar I have the same question. Did one of you find an answer?

**Comment by casparcaspar** [comment]
Yes, I did. I'm training in stateless mode and later use the weights with model.get_weights() and model.set_weights() in a stateful model with timesteps=1 and batchsize=1 for prediction.

**Comment by casparcaspar** [comment]
@JustAnotherUsername123 I've worked on the issue quite intense for my thesis. What are you planing to do with your model?

**Comment by rpicatoste** [comment]
Hi, I made a function to convert a model used during training to a model meant to be used in inference, receiving just one sample and one time step at a time. In case it helps:
[Gist with the function](https://gist.github.com/rpicatoste/02cecac1ed52524301e3ab423dac888b)

It basically copies the model, changes 'stateful' to True and the input shape to (1, 1, original_value), and transfers the weights from the trained model.

PD: I didn't find a better solution to this problem, which I thought it would be a very common one. Please let me know any comments or solutions that I missed.

**Comment by falaktheoptimist** [comment]
> Hi, I made a function to convert a model used during training to a model meant to be used in inference, receiving a single sample with any number of time steps (so a single time step sample can be used for real-time use). In case it helps:
> [Gist with the function](https://gist.github.com/rpicatoste/02cecac1ed52524301e3ab423dac888b)
> 
> It basically copies the model, changes 'stateful' to True and the input shape to (1, None, original_value), and transfers the weights from the trained model.
> 
> PS: I didn't find a better solution to this problem, which I thought it would be a very common one. Please let me know any comments or solutions that I missed.

Thank you @rpicatoste - I didn't know that all the layers had a stateful tag and all of those had to be reset. I had initially cloned the model and just changed the stateful flag in LSTM to true.

**Comment by rpicatoste** [comment]
@falaktheoptimist I'm glad if it helped! This is the solution I am still using (although I didn't search again). 

Actually, not all layers have the stateful parameters in their config, but if you have several RNN layers, you will have to set it in each one of those. Hence the check for the parameter in each layer.
